{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pipeline: Gene-Drug association for CCLE cancer cell line pharmacogenomic data\n",
    "\n",
    "Find significance of gene-drug association for an interested gene list with gene cnv, gene expression, drug response and dependency data, and box plot of drug/dep for cnv-status/gene-expression for the interested genes.\n",
    "\n",
    "Updated August 15th, 2019\n",
    "\n",
    "@Copyright Zi-Ming Zhao ziming.gt@gmail.com\n",
    "\n",
    "## Overall summary:\n",
    "The pipeline is evaluating the drug effects of selected gene’s CNA changes by using the CCLE cell line genomic and drug data (CTRP v2) [Viswanathan et al. (Nature, 2017); Adams et al. (ACS Chem Biol, 2014)]. \n",
    "\n",
    "The Cancer Cell Line Encyclopedia (CCLE) cell line genomic and drug data (CTRP v2) were used [Viswanathan et al. (Nature, 2017); Adams et al. (ACS Chem Biol, 2014)]. The CCLE drug response data were downloaded from Cancer Therapeutics Response Portal (www.broadinstitute.org/ctrp), and CCLE gene-level CNA and gene expression data from depMap data portal (‘public_19Q1_gene_cn.csv’ and ‘CCLE_depMap_19Q1_TPM.csv’ https://depmap.org/portal/download/) [citations:\n",
    "Cancer Cell Line Encyclopedia Consortium, and Genomics of Drug Sensitivity in Cancer Consortium. 2015. Pharmacogenomic Agreement between Two Cancer Cell Line Data Sets. Nature 528 (7580):84–87. https://doi.org/10.1038/nature15736.\n",
    "Jordi Barretina, Giordano Caponigro, Nicolas Stransky, Kavitha Venkatesan, William R. Sellers, Robert Schlegel, Levi A. Garraway, et. al. 2012. The Cancer Cell Line Encyclopedia Enables Predictive Modelling of Anticancer Drug Sensitivity. Nature 483 (7391):603–7. https://doi.org/10.1038/nature11003.]. \n",
    "\n",
    "For CCLE drug data, area-under-concentration-response curve (AUC) sensitivity scores were used for each cancer cell line and each drug. In total, I collected gene-level CNA data from 668 CCLE cell lines, with a total of 545 drugs tested. With the CCLE gene-level CNA and AUC drug sensitivity scores, I performed drug effects functional analyses for selected  genes. Particularly, I took the selected gene list, calculated P-value from the Pearson correlation between each gene’s CNA and each drug across cell lines for the interested tumor types or pan-cancer, and further calculated Q-value by multiple testing Bonferroni correction for the P-values. Significant gene-CNA and drug associations were kept (Q-value > 0.1), and is further evaluated for the gene-expression and drug associations. Final significant genes were presented if the gene-drug association is significant for both CNA and gene expression in the same direction, e.g. drug sensitivity with both high CNA and high gene expression.\n",
    "\n",
    "### Required input files/data:\n",
    "CCLE_gene_shared_drug.csv #file with drug response auc for genes CNV for each cell line\n",
    "CCLE_gene_shared_deps.csv file with dependency for gene CNV for each cell lines;\n",
    "cnv_drug_sharedCellLines.csv ##file with IDs matching between drug/dependency cell lines and ccle cell lines\n",
    "drug_target.csv\n",
    "\n",
    "cnv_ctrpv2_gene_drug_tumortype_correlation.csv\n",
    "gex_ctrpv2_gene_drug_tumortype_correlation.csv\n",
    "\n",
    "### Output files:\n",
    "Gene cnv/gene-expression and drug/dependency association.\n",
    "Box plot of dep/drugs for cnv-status/gene-expression for the interested genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import json\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import statistics as st\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy\n",
    "from scipy.stats.stats import pearsonr\n",
    "from matplotlib_venn import venn3\n",
    "from matplotlib_venn import venn2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Missing Genes:\n",
      "GOLGA6L6\n",
      "POTEB\n",
      "POTEB2\n",
      "total_gene_number\t12\texist_gene_number\t9\tMissing_gene_number\t3\n",
      "Missing Genes:\n",
      "GOLGA6L6\n",
      "HLA-DQA1\n",
      "HLA-DQB1\n",
      "HLA-DRB1\n",
      "HLA-DRB5\n",
      "MACROD2\n",
      "OR4M2\n",
      "OR4N4\n",
      "POTEB\n",
      "POTEB2\n",
      "RBFOX1\n",
      "TPTE\n",
      "total_gene_number\t12\texist_gene_number\t0\tMissing_gene_number\t12\n",
      "Missing Genes:\n",
      "GOLGA6L6\n",
      "HLA-DQA1\n",
      "HLA-DQB1\n",
      "HLA-DRB1\n",
      "HLA-DRB5\n",
      "MACROD2\n",
      "OR4M2\n",
      "OR4N4\n",
      "POTEB\n",
      "POTEB2\n",
      "RBFOX1\n",
      "TPTE\n",
      "total_gene_number\t12\texist_gene_number\t0\tMissing_gene_number\t12\n",
      "Missing Genes:\n",
      "GOLGA6L6\n",
      "MACROD2\n",
      "OR4M2\n",
      "OR4N4\n",
      "POTEB\n",
      "POTEB2\n",
      "total_gene_number\t12\texist_gene_number\t6\tMissing_gene_number\t6\n",
      "Missing Genes:\n",
      "GOLGA6L6\n",
      "MACROD2\n",
      "OR4M2\n",
      "OR4N4\n",
      "POTEB\n",
      "POTEB2\n",
      "total_gene_number\t12\texist_gene_number\t6\tMissing_gene_number\t6\n"
     ]
    }
   ],
   "source": [
    "### Check the genes' presence and absence in the datasets\n",
    "\n",
    "genes=['GOLGA6L6','HLA-DQA1','HLA-DQB1','HLA-DRB1','HLA-DRB5','MACROD2','OR4M2','OR4N4','POTEB','POTEB2','RBFOX1','TPTE']\n",
    "\n",
    "g1=ExistGenesWithoutCounts(genes,ccle_genes) \n",
    "g2=ExistGenesWithoutCounts(genes,ccle_gex_genes) \n",
    "g3=ExistGenesWithoutCounts(genes,ccle_cnv_genes) \n",
    "g4=ExistGenesWithoutCounts(genes,dep_ctrpv2_genes) \n",
    "g5=ExistGenesWithoutCounts(genes,drug_ctrpv2_genes) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genes tested: \n",
      "6\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/zhaoz/anaconda3/lib/python3.6/site-packages/scipy/stats/stats.py:3005: RuntimeWarning: Mean of empty slice.\n",
      "  mx = x.mean()\n",
      "/Users/zhaoz/anaconda3/lib/python3.6/site-packages/numpy/core/_methods.py:80: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ret = ret.dtype.type(ret / rcount)\n",
      "/Users/zhaoz/anaconda3/lib/python3.6/site-packages/scipy/stats/stats.py:3006: RuntimeWarning: Mean of empty slice.\n",
      "  my = y.mean()\n",
      "/Users/zhaoz/anaconda3/lib/python3.6/site-packages/scipy/stats/stats.py:3010: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  r = r_num / r_den\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Empty output! Nothing came out to be significant!\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "335.5937530994415"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get starting time\n",
    "\n",
    "import time\n",
    "start = time.time()\n",
    "#genes=['GOLGA6L6','HLA-DQA1','HLA-DQB1','HLA-DRB1','HLA-DRB5','MACROD2','OR4M2','OR4N4','POTEB','POTEB2','RBFOX1','TPTE']\n",
    "genes=['HLA-DQA1','HLA-DQB1','HLA-DRB1','HLA-DRB5','RBFOX1','TPTE']\n",
    "\n",
    "output_f=SignificantGenes_cnv_drug(genes) # 5 min for 6 genes\n",
    "\n",
    "#output_f=output_f[output_f['gex_r']!=0]\n",
    "output_f.to_csv('Validated_Genes_DrugEffects_GeX.csv')\n",
    "\n",
    "elapsed_time_fl = (time.time() - start) \n",
    "elapsed_time_fl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drug</th>\n",
       "      <th>gene_symbol_of_protein_target</th>\n",
       "      <th>tumor_type</th>\n",
       "      <th>Q_value</th>\n",
       "      <th>p_value</th>\n",
       "      <th>correlation_r</th>\n",
       "      <th>data_size</th>\n",
       "      <th>gex_dir</th>\n",
       "      <th>gex_r</th>\n",
       "      <th>gex_z</th>\n",
       "      <th>gex_num_cells</th>\n",
       "      <th>cnv_dir</th>\n",
       "      <th>cnv_r</th>\n",
       "      <th>cnv_z</th>\n",
       "      <th>cnv_num_cells</th>\n",
       "      <th>Q_value_dep</th>\n",
       "      <th>p_value_dep</th>\n",
       "      <th>correlation_r_dep</th>\n",
       "      <th>data_size_dep</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RBFOX1</th>\n",
       "      <td>SKI-II</td>\n",
       "      <td>SPHK1</td>\n",
       "      <td>PanCancer</td>\n",
       "      <td>1.752631</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>-0.167637</td>\n",
       "      <td>423.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.918196</td>\n",
       "      <td>0.319699</td>\n",
       "      <td>0.047013</td>\n",
       "      <td>450.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TPTE</th>\n",
       "      <td>CAY10576</td>\n",
       "      <td>IKBKE</td>\n",
       "      <td>PanCancer</td>\n",
       "      <td>1.954234</td>\n",
       "      <td>0.000598</td>\n",
       "      <td>0.167614</td>\n",
       "      <td>416.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.932527</td>\n",
       "      <td>0.822088</td>\n",
       "      <td>-0.010629</td>\n",
       "      <td>450.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RBFOX1</th>\n",
       "      <td>fluorouracil</td>\n",
       "      <td>TYMS</td>\n",
       "      <td>PanCancer</td>\n",
       "      <td>2.373052</td>\n",
       "      <td>0.000726</td>\n",
       "      <td>-0.160350</td>\n",
       "      <td>441.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.918196</td>\n",
       "      <td>0.319699</td>\n",
       "      <td>0.047013</td>\n",
       "      <td>450.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RBFOX1</th>\n",
       "      <td>BMS-345541</td>\n",
       "      <td>IKBKB</td>\n",
       "      <td>PanCancer</td>\n",
       "      <td>8.132884</td>\n",
       "      <td>0.002487</td>\n",
       "      <td>-0.146541</td>\n",
       "      <td>424.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.918196</td>\n",
       "      <td>0.319699</td>\n",
       "      <td>0.047013</td>\n",
       "      <td>450.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RBFOX1</th>\n",
       "      <td>BRD-K07442505</td>\n",
       "      <td>BAX</td>\n",
       "      <td>PanCancer</td>\n",
       "      <td>8.253115</td>\n",
       "      <td>0.002524</td>\n",
       "      <td>-0.143332</td>\n",
       "      <td>442.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.918196</td>\n",
       "      <td>0.319699</td>\n",
       "      <td>0.047013</td>\n",
       "      <td>450.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 drug gene_symbol_of_protein_target tumor_type   Q_value  \\\n",
       "gene                                                                       \n",
       "RBFOX1         SKI-II                         SPHK1  PanCancer  1.752631   \n",
       "TPTE         CAY10576                         IKBKE  PanCancer  1.954234   \n",
       "RBFOX1   fluorouracil                          TYMS  PanCancer  2.373052   \n",
       "RBFOX1     BMS-345541                         IKBKB  PanCancer  8.132884   \n",
       "RBFOX1  BRD-K07442505                           BAX  PanCancer  8.253115   \n",
       "\n",
       "         p_value  correlation_r  data_size  gex_dir  gex_r  gex_z  \\\n",
       "gene                                                                \n",
       "RBFOX1  0.000536      -0.167637      423.0      0.0    0.0    0.0   \n",
       "TPTE    0.000598       0.167614      416.0      0.0    0.0    0.0   \n",
       "RBFOX1  0.000726      -0.160350      441.0      0.0    0.0    0.0   \n",
       "RBFOX1  0.002487      -0.146541      424.0      0.0    0.0    0.0   \n",
       "RBFOX1  0.002524      -0.143332      442.0      0.0    0.0    0.0   \n",
       "\n",
       "        gex_num_cells  cnv_dir  cnv_r  cnv_z  cnv_num_cells  Q_value_dep  \\\n",
       "gene                                                                       \n",
       "RBFOX1            0.0      0.0    0.0    0.0            0.0     1.918196   \n",
       "TPTE              0.0      0.0    0.0    0.0            0.0     4.932527   \n",
       "RBFOX1            0.0      0.0    0.0    0.0            0.0     1.918196   \n",
       "RBFOX1            0.0      0.0    0.0    0.0            0.0     1.918196   \n",
       "RBFOX1            0.0      0.0    0.0    0.0            0.0     1.918196   \n",
       "\n",
       "        p_value_dep  correlation_r_dep  data_size_dep  \n",
       "gene                                                   \n",
       "RBFOX1     0.319699           0.047013          450.0  \n",
       "TPTE       0.822088          -0.010629          450.0  \n",
       "RBFOX1     0.319699           0.047013          450.0  \n",
       "RBFOX1     0.319699           0.047013          450.0  \n",
       "RBFOX1     0.319699           0.047013          450.0  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "output_f.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/zhaoz/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2717: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "545"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Open cell line gene cnv file:\n",
    "cnv_f = '/Users/zhaoz/Dropbox/gene_drug_pipeline/CCLE_gene_shared_drug.csv'\n",
    "cnv = pd.read_csv(cnv_f,sep=',',index_col=0)\n",
    "cnv.head() #(668, 16341)\n",
    "# File 'cnv_drug_sharedCellLines.csv' with compound, target, AUC, and shared cell lines with cnv files.\n",
    "file_shared='/Users/zhaoz/Dropbox/gene_drug_pipeline/cnv_drug_sharedCellLines.csv'\n",
    "cnv_drug= pd.read_csv(file_shared,sep=',',index_col=0)\n",
    "cnv_drug.head()\n",
    "dep_f = '/Users/zhaoz/Dropbox/gene_drug_pipeline/CCLE_gene_shared_deps.csv'\n",
    "dep = pd.read_csv(dep_f,sep=',',index_col=0)\n",
    "dep.head() #16341 columns\n",
    "cell_line=pd.concat([cnv_drug['CCLE_Name.1'],cnv_drug['ccle_primary_site']],axis=1)\n",
    "cell_lines=cell_line.drop_duplicates()\n",
    "dep_mod=pd.concat([cell_lines,dep],axis=1,join='inner')\n",
    "dep_mod.head()\n",
    "\n",
    "# genes in dependency and drug data\n",
    "dep_ctrpv2_genes=dep_mod.T[10:]\n",
    "dep_ctrpv2_genes=dep_ctrpv2_genes.index\n",
    "dep_ctrpv2_genes\n",
    "drug_ctrpv2_genes=cnv.T[8:]\n",
    "drug_ctrpv2_genes=drug_ctrpv2_genes.index\n",
    "drug_ctrpv2_genes\n",
    "\n",
    "# Open file drug targeted gene \"drug_target.csv\"\n",
    "file_drug_target='/Users/zhaoz/Dropbox/gene_drug_pipeline/drug_target.csv'\n",
    "drug_target= pd.read_csv(file_drug_target,sep=',',index_col=0)\n",
    "drug_target.index=drug_target['cpd_name']\n",
    "drug_target.head()\n",
    "\n",
    "### CCLE validation data: gene expression and cnv vs drug\n",
    "# Open cell line gene cnv file:\n",
    "f_cnv = '/Users/zhaoz/Dropbox/gene_drug_pipeline/cnv_ctrpv2_gene_drug_tumortype_correlation.csv'\n",
    "cnv_corr = pd.read_csv(f_cnv,sep=',',index_col=0)\n",
    "cnv_corr['context']=cnv_corr['context'].str.replace('all_CCLs','PanCancer')\n",
    "cnv_corr.head() #(195214, 8)\n",
    "# Open cell line gene expression file:\n",
    "f_gex = '/Users/zhaoz/Dropbox/gene_drug_pipeline/gex_ctrpv2_gene_drug_tumortype_correlation.csv'\n",
    "gex = pd.read_csv(f_gex,sep=',',index_col=0)\n",
    "gex['context']=gex['context'].str.replace('all_CCLs','PanCancer')\n",
    "\n",
    "gex.head() #(1184225, 8)\n",
    "# genes found in the ccle gex and cnv data\n",
    "ccle_gex_genes=gex['hugo_gene_symbol']\n",
    "ccle_cnv_genes=cnv_corr['hugo_gene_symbol']\n",
    "ccle_genes=ccle_gex_genes.append(ccle_cnv_genes)\n",
    "ccle_genes=ccle_genes.unique() #16989\n",
    "\n",
    "drugs=cnv_drug['cpd_name'].unique() #545 drugs including dosing difference\n",
    "num_drugs=len(drugs)\n",
    "interested_tumor_types=['PanCancer']\n",
    "num_tumorTypes=len(interested_tumor_types)\n",
    "num_tumorTypes #1\n",
    "num_drugs #545"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# cnv-drug q value cutoff\n",
    "def SignificantGenes_cnv_drug(gene_list):\n",
    "    test_genes=gene_list\n",
    "    test_drugs=drugs\n",
    "    num_genes=len(test_genes)\n",
    "    TotalTests=num_genes*num_drugs*num_tumorTypes\n",
    "    TotalTests_dep=num_genes*num_tumorTypes\n",
    "    \n",
    "    Gene_drug_tumor=pd.DataFrame()\n",
    "    df=pd.DataFrame()\n",
    "    cnv_cutoff=0.2\n",
    "    p_value_cutoff=0.05\n",
    "    q_value_cutoff=0.1\n",
    "    min_data=5;\n",
    "    min_cutoff=min_data-1;\n",
    "    gene_names=''\n",
    "    i=0\n",
    "    count=0\n",
    "    print (\"Number of genes tested: \")\n",
    "    print (num_genes)\n",
    "    for gene in test_genes:\n",
    "        i=i+1\n",
    "        gene_names=gene+'_'+gene_names\n",
    "        order=str(i)+':'+gene\n",
    "        #print (order)\n",
    "        for tumor in interested_tumor_types: \n",
    "            (tmp_r_dep,tmp_p_dep,data_size_dep)=Gene_CNV_dep_TumorType(gene,tumor)\n",
    "            tmp_q_dep=tmp_p_dep*TotalTests_dep       \n",
    "            for drug in test_drugs:\n",
    "                (tmp_r,tmp_p,data_size)=Gene_CNV_drug_TumorType(gene,drug,tumor,cnv_cutoff)           \n",
    "                tmp_q=tmp_p*TotalTests\n",
    "                if (tmp_q<q_value_cutoff):\n",
    "                    count=count+1\n",
    "                    #gene_count=ExsitGenes_WithCounts[gene]                    \n",
    "                if (tumor=='PanCancer'):\n",
    "                    validate_tumor='all_CCLs'\n",
    "                else:\n",
    "                    validate_tumor=tumor\n",
    "\n",
    "                (cnv_direction,cnv_peason_corr, cnv_zscore,cnv_num_cells)=find_gene_gex(gene,validate_tumor,drug,cnv_corr)\n",
    "                (gex_direction,gex_peason_corr, gex_zscore,gex_num_cells)=find_gene_gex(gene,validate_tumor,drug,gex)\n",
    "                df = df.append({'gene':gene, 'drug':drug,'tumor_type':tumor,'correlation_r':tmp_r,'p_value':tmp_p,'Q_value':tmp_q,'data_size':data_size,'gex_dir':gex_direction,'gex_r':gex_peason_corr, 'gex_z':gex_zscore,'gex_num_cells':gex_num_cells,'cnv_dir':cnv_direction,'cnv_r':cnv_peason_corr, 'cnv_z':cnv_zscore,'cnv_num_cells':cnv_num_cells,'correlation_r_dep':tmp_r_dep,'p_value_dep':tmp_p_dep,'Q_value_dep':tmp_q_dep,'data_size_dep':data_size_dep}, ignore_index=True)\n",
    "\n",
    "    df.index=df['drug']\n",
    "    output_content=pd.concat([df,drug_target],axis=1,join='inner') \n",
    "    output_content.index=output_content['gene']\n",
    "\n",
    "    # re-organize the columns\n",
    "    dc=output_content\n",
    "    Gene_drug_tumor=pd.concat([dc['drug'],dc['gene_symbol_of_protein_target'],dc['tumor_type'],dc['Q_value'],dc['p_value'],dc['correlation_r'],dc['data_size'],dc['gex_dir'],dc['gex_r'],dc['gex_z'],dc['gex_num_cells'],dc['cnv_dir'],dc['cnv_r'],dc['cnv_z'],dc['cnv_num_cells'],dc['Q_value_dep'],dc['p_value_dep'],dc['correlation_r_dep'],dc['data_size_dep']],axis=1,join='inner')\n",
    "    Gene_drug_tumor.index=dc['gene']\n",
    "    # rank by p_value\n",
    "    Gene_drug_tumor.sort_values(\"Q_value\", inplace = True) \n",
    "    size=Gene_drug_tumor.shape[0]\n",
    "        \n",
    "    if (count>0): # save the file if it is not empty          \n",
    "        return (Gene_drug_tumor)\n",
    "\n",
    "    else:\n",
    "        print (\"Nothing came out to be significant!\")\n",
    "        return (Gene_drug_tumor)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Main function to take the gene name, drug name, tumor time, and cnv_cutoff as input files;\n",
    "#### and output the box plot and pearson correlation r and p values\n",
    "### For pan-cancer, use TumorType as All\n",
    "\n",
    "def Gene_CNV_drug_TumorType(GeneID,drug,TumorType,cnv_cutoff):\n",
    "    import seaborn as sns\n",
    "    import matplotlib.pyplot as plt\n",
    "    if (TumorType == \"PanCancer\"):\n",
    "        drug_tumor=cnv_drug\n",
    "    else:\n",
    "        drug_tumor=cnv_drug[cnv_drug['ccle_primary_site']==TumorType]    \n",
    "    gene_cnv=cnv[GeneID]\n",
    "    gene_cnv=gene_cnv.rename(columns={GeneID:'cnv'}, inplace=True)\n",
    "    gene_drug=drug_tumor[drug_tumor['cpd_name']==drug] \n",
    "    gene_cnv_drug=pd.concat([gene_drug,gene_cnv],axis=1,join='inner') \n",
    "    rows=gene_cnv_drug.shape[0]    \n",
    "    # too few drug data or tumor type data, skip it. size of gene_cnv_drug\n",
    "    if rows<3:        \n",
    "        return (0,1.0,int(rows)) # if not enough data, force r=0 and p=1\n",
    "    else:        \n",
    "        gene_cnv_drug.columns.values[-1] = \"cnv\"\n",
    "        gene_cnv_drug.columns.values[0] = \"AUC\"\n",
    "        gene_cnv_drug=gene_cnv_drug.sort_values(by='cnv', ascending=True)\n",
    "        gene_cnv_s=gene_cnv_drug['cnv']\n",
    "        gain_cnv = gene_cnv_s > cnv_cutoff; #default 0.2\n",
    "        gain_cnv\n",
    "        gain_cnv.sum() #10 pts\n",
    "        gene_cnv_drug_sort=pd.concat([gene_cnv_drug,gain_cnv],axis=1,join='inner') \n",
    "        gene_cnv_drug_sort.columns.values[-1] = 'CNV_status'\n",
    "        gene_cnv_drug=gene_cnv_drug.dropna()\n",
    "\n",
    "        (r,p)=pearsonr(gene_cnv_drug['cnv'],gene_cnv_drug['AUC'])\n",
    "        return (r,p,int(rows))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Create a function to take the gene name as the input and output the box plot\n",
    "\n",
    "def Gene_CNV_dep_TumorType(GeneID,TumorType):\n",
    "    import seaborn as sns\n",
    "    import matplotlib.pyplot as plt\n",
    "    if (TumorType == \"PanCancer\"):\n",
    "        dep_tumor=dep_mod\n",
    "    else:\n",
    "        dep_tumor=dep_mod[dep_mod['ccle_primary_site']==TumorType]    \n",
    "    gene_cnv=cnv[GeneID]\n",
    "    gene_cnv=gene_cnv.rename(columns={GeneID:'cnv'}, inplace=True)\n",
    "    gene_dep=dep_tumor[GeneID]\n",
    "    gene_cnv_dep=pd.concat([gene_cnv,gene_dep],axis=1,join='inner') \n",
    "    gene_cnv_dep.columns.values[0] = \"cnv\"\n",
    "    gene_cnv_dep.columns.values[1] = \"dep\"\n",
    "    gene_cnv_dep=gene_cnv_dep.sort_values(by='cnv', ascending=True)\n",
    "    gene_cnv_dep=gene_cnv_dep.dropna()\n",
    "    (r,p)=pearsonr(gene_cnv_dep['cnv'],gene_cnv_dep['dep'])\n",
    "    cellLineNum=gene_cnv_dep.shape[0]\n",
    "    return (r,p,int(cellLineNum))   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Function to validate the gene, drug, tumor_type and data \n",
    "def find_gene_gex(gene,tumor_type,drug,data):\n",
    "    g=gene\n",
    "    t=tumor_type\n",
    "    d=drug\n",
    "    f=data\n",
    "    f_t=f[f['context']==t]\n",
    "    f_t_g=f_t[f_t['hugo_gene_symbol']==g]\n",
    "    indx=f_t_g.index\n",
    "    if (indx.contains(d)):\n",
    "        tmp=f_t_g.loc[d,:]\n",
    "        \n",
    "        return (tmp['correl_direction'],tmp['pearson_correlation'],tmp['correlation_zscore'],tmp['number_of_cell_lines'])\n",
    "    else:\n",
    "        return (0,0,0,0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Function to validate the gene, drug,and data for a fixed tumor type\n",
    "def find_gene_drug(gene,drug,data):\n",
    "    g=gene\n",
    "    d=drug\n",
    "    f=data\n",
    "    indx=f.index\n",
    "    if (indx.contains(d)):\n",
    "        tmp=f.loc[d,:]        \n",
    "        return (tmp['correl_direction'],tmp['pearson_correlation'],tmp['correlation_zscore'],tmp['number_of_cell_lines'])\n",
    "    else:\n",
    "        return (0,0,0,0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to get exist genes from a query gene list [e.g. co-exisiting ccle genes]\n",
    "# input: test_gene_with_counts (dtype: series), query_gene_list\n",
    "# output: exist_gene_with_counts\n",
    "# Example: ExistGenesWithCounts(PTP0_all,ccle_genes); ExistGenesWithCounts(PTP0_all,cnv)\n",
    "def ExistGenesWithoutCounts(test_gene_without_counts,query_gene_list):\n",
    "    test_genes=test_gene_without_counts\n",
    "    #print (test_genes)\n",
    "    Exist_genes=[]\n",
    "    NotExist=[]\n",
    "    print (\"Missing Genes:\")\n",
    "    for gene in test_genes:\n",
    "        #print (gene)\n",
    "        if gene in query_gene_list:\n",
    "            Exist_genes.append(gene)\n",
    "        else:\n",
    "            NotExist.append(gene)\n",
    "            print (gene)\n",
    "    p1='total_gene_number\\t'+str(len(test_genes))+\"\\texist_gene_number\\t\"+str(len(Exist_genes))+\"\\tMissing_gene_number\\t\"+str(len(NotExist))\n",
    "    print (p1)\n",
    "    #print (NotExist)\n",
    "\n",
    "    #rowData = dfObj.loc[ ['c' , 'b'] , : ]\n",
    "    return (Exist_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Create a function to take the gene name as the input and output the box plot\n",
    "# Separate the data by GeneID1, the status of CNV, and look at gene1's CNV status's impact on gene2's genetic dependency\n",
    "\n",
    "def Gene_CNV_dep_plotting(GeneID1_CNV,GeneID2_dep,TumorType,cnv_cutoff):\n",
    "    import seaborn as sns\n",
    "    import matplotlib.pyplot as plt\n",
    "    if (TumorType == \"PanCancer\"):\n",
    "        dep_tumor=dep_mod\n",
    "    else:\n",
    "        dep_tumor=dep_mod[dep_mod['ccle_primary_site']==TumorType]    \n",
    "    gene_cnv=cnv[GeneID1_CNV]\n",
    "    x_axis=GeneID1_CNV+'_cnv'\n",
    "    gene_cnv=gene_cnv.rename(columns={GeneID1_CNV:x_axis}, inplace=True)\n",
    "    gene_dep=dep_tumor[GeneID2_dep]\n",
    "    gene_cnv_dep=pd.concat([gene_cnv,gene_dep],axis=1,join='inner') \n",
    "    \n",
    "    y_axis=GeneID2_dep+'_genetic_dependency'\n",
    "    gene_cnv_dep.columns.values[0] = x_axis\n",
    "    gene_cnv_dep.columns.values[1] = 'genetic_dependency'\n",
    "    gene_cnv_dep=gene_cnv_dep.dropna()\n",
    "    (r,p)=pearsonr(gene_cnv_dep[x_axis],gene_cnv_dep['genetic_dependency'])\n",
    "    cellLineNum=gene_cnv_dep.shape[0]\n",
    "\n",
    "    \n",
    "    gene_cnv_dep=gene_cnv_dep.sort_values(by=x_axis, ascending=True)\n",
    "    gene_cnv_s=gene_cnv_dep[x_axis]\n",
    "    gain_cnv = gene_cnv_s > cnv_cutoff; #default 0.2\n",
    "    gain_cnv\n",
    "    gain_cnv.sum() #10 pts\n",
    "    gene_cnv_dep_sort=pd.concat([gene_cnv_dep,gain_cnv],axis=1,join='inner') \n",
    "    \n",
    "\n",
    "    gene_cnv_dep_sort.columns.values[2] = 'cnv_status'\n",
    "    gene_cnv_dep_sort.head()\n",
    "    gene_cnv_dep_sort=gene_cnv_dep_sort.dropna()\n",
    "    x=gene_cnv_dep_sort['cnv_status']\n",
    "    y=gene_cnv_dep_sort['genetic_dependency']\n",
    "    order=['NoGain','Gain']\n",
    "    df=gene_cnv_dep_sort\n",
    "    ax = sns.boxplot(x=x,y=y)\n",
    "    \n",
    "    TumorType=TumorType.replace('/','')\n",
    "    TumorType=TumorType.replace(' ','')\n",
    "    fig, ax = plt.subplots(1, 1)\n",
    "    plt.xlim(-2,2)\n",
    "    ax.plot(gene_cnv_dep[x_axis],gene_cnv_dep['genetic_dependency'], '.')\n",
    "    ax.set_xlabel(x_axis)\n",
    "    ax.set_ylabel(y_axis)\n",
    "\n",
    "    return (r,p,int(cellLineNum))  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Test plotting function to take the gene name, drug name, tumor time, and cnv_cutoff as input files;\n",
    "#### and output the box plot and pearson correlation r and p values\n",
    "### For pan-cancer, use TumorType as All\n",
    "\n",
    "def Gene_CNV_drug_TumorType_plotting(GeneID,drug,TumorType,cnv_cutoff):\n",
    "    import seaborn as sns\n",
    "    import matplotlib.pyplot as plt\n",
    "    if (TumorType == \"PanCancer\"):\n",
    "        drug_tumor=cnv_drug\n",
    "    else:\n",
    "        drug_tumor=cnv_drug[cnv_drug['ccle_primary_site']==TumorType]    \n",
    "    gene_cnv=cnv[GeneID]\n",
    "    gene_cnv=gene_cnv.rename(columns={GeneID:'cnv'}, inplace=True)\n",
    "    gene_drug=drug_tumor[drug_tumor['cpd_name']==drug]    \n",
    "    \n",
    "    # fix bug: two few drug data or tumor type data, skip it. size of gene_cnv_drug\n",
    "    # Save data size\n",
    "    gene_cnv_drug=pd.concat([gene_drug,gene_cnv],axis=1,join='inner') \n",
    "    rows=gene_cnv_drug.shape[0]\n",
    "\n",
    "    if rows<3:\n",
    "        print (GeneID+\" \"+drug+\" \"+TumorType+\" Not enough data!\")\n",
    "        return (0,1,int(rows)) # if not enough data, force r=0 and p=1\n",
    "    else:\n",
    "        \n",
    "        gene_cnv_drug.columns.values[-1] = \"cnv\"\n",
    "        gene_cnv_drug.columns.values[0] = \"AUC\"\n",
    "        gene_cnv_drug=gene_cnv_drug.sort_values(by='cnv', ascending=True)\n",
    "        gene_cnv_s=gene_cnv_drug['cnv']\n",
    "        gain_cnv = gene_cnv_s > cnv_cutoff; #default 0.2\n",
    "        gain_cnv\n",
    "        gain_cnv.sum() #10 pts\n",
    "        gene_cnv_drug_sort=pd.concat([gene_cnv_drug,gain_cnv],axis=1,join='inner') \n",
    "        gene_cnv_drug_sort.columns.values[-1] = 'CNA_status'\n",
    "        gene_cnv_drug=gene_cnv_drug.dropna()\n",
    "        (r,p)=pearsonr(gene_cnv_drug['cnv'],gene_cnv_drug['AUC'])\n",
    "        print (r,p)\n",
    "        x_label=GeneID+'_CNA_status'\n",
    "        y_label=drug+'_AUC'\n",
    "        ax = sns.boxplot(x=gene_cnv_drug_sort['CNA_status'], y=gene_cnv_drug_sort['AUC'])\n",
    "        order=['NoGain','Gain']\n",
    "        TumorType=TumorType.replace('/','')\n",
    "        TumorType=TumorType.replace(' ','')\n",
    "\n",
    "\n",
    "        fig, ax = plt.subplots(1, 1)\n",
    "        plt.xlim(-2,2)\n",
    "        ax.plot(gene_cnv_drug['cnv'],gene_cnv_drug['AUC'], '.')\n",
    "        ax.set_xlabel(x_label)\n",
    "        ax.set_ylabel(y_label)\n",
    "        \n",
    "        file_name=GeneID+\"_\"+drug+\"_\"+TumorType+'_cna_cutoff_'+str(cnv_cutoff)+'_gain_cna_drug.csv'\n",
    "        #gene_cnv_drug_sort['CNA_status']=gene_cnv_drug_sort['CNA_status'].apply(ChangeTrueFalse)\n",
    "        gene_cnv_drug_sort.to_csv(file_name)\n",
    "        \n",
    "        return (r,p,int(rows))\n",
    "\n",
    "    # Add gene name and p-value to the box plot"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
